addons/utm/models/utm_mixin.py

```diff
@@ -59,3 +53,37 @@ def tracking_fields(self):
             ('utm_source', 'source_id', 'odoo_utm_source'),
             ('utm_medium', 'medium_id', 'odoo_utm_medium'),
         ]
+
+    def _generate_url_parameters(self):
+        """Generate the URL parameters for the UTM records.
+
+        Generate the parameters to add in the URL to be able to track the UTM records
+        (source, medium, campaign...).
+        """
+        self.ensure_one()
+        url_parameters = {}
+        for url_parameter, field_name, cookie_name in self.env['utm.mixin'].tracking_fields():
+            record = getattr(self, field_name)
+            if record:
+                url_parameters[url_parameter] = record.identifier
+        return url_parameters
+
+    def _find_or_create_record(self, model_name, identifier):
```

Comment:
Not sure this method adds readability to the code ... did you change anything ?

Reply:
The big change is that now we search first based on the identifier, but splitting it in a separate method help me to fully test it

Reply:
Ok